In [0]:
import pandas as pd

In [0]:
df = pd.read_csv("pokemon_data.csv")
df.drop(columns = ['#'], inplace = True)
df.head()

Name 
 Type 1 
 Type 2 
 HP 
 Attack 
 Defense 
 Sp. Atk 
 Sp. Def 
 Speed 
 Generation 
 Legendary 
 
 
 
 
 0 
 Bulbasaur 
 Grass 
 Poison 
 45 
 49 
 49 
 65 
 65 
 45 
 1 
 False 
 
 
 1 
 Ivysaur 
 Grass 
 Poison 
 60 
 62 
 63 
 80 
 80 
 60 
 1 
 False 
 
 
 2 
 Venusaur 
 Grass 
 Poison 
 80 
 82 
 83 
 100 
 100 
 80 
 1 
 False 
 
 
 3 
 VenusaurMega Venusaur 
 Grass 
 Poison 
 80 
 100 
 123 
 122 
 120 
 80 
 1 
 False 
 
 
 4 
 Charmander 
 Fire 
 NaN 
 39 
 52 
 43 
 60 
 50 
 65 
 1 
 False

In [0]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Name        800 non-null    object
 1   Type 1      800 non-null    object
 2   Type 2      414 non-null    object
 3   HP          800 non-null    int64 
 4   Attack      800 non-null    int64 
 5   Defense     800 non-null    int64 
 6   Sp. Atk     800 non-null    int64 
 7   Sp. Def     800 non-null    int64 
 8   Speed       800 non-null    int64 
 9   Generation  800 non-null    int64 
 10  Legendary   800 non-null    bool  
dtypes: bool(1), int64(7), object(3)
memory usage: 63.4+ KB


In [0]:
df.describe()

HP 
 Attack 
 Defense 
 Sp. Atk 
 Sp. Def 
 Speed 
 Generation 
 
 
 
 
 count 
 800.000000 
 800.000000 
 800.000000 
 800.000000 
 800.000000 
 800.000000 
 800.00000 
 
 
 mean 
 69.258750 
 79.001250 
 73.842500 
 72.820000 
 71.902500 
 68.277500 
 3.32375 
 
 
 std 
 25.534669 
 32.457366 
 31.183501 
 32.722294 
 27.828916 
 29.060474 
 1.66129 
 
 
 min 
 1.000000 
 5.000000 
 5.000000 
 10.000000 
 20.000000 
 5.000000 
 1.00000 
 
 
 25% 
 50.000000 
 55.000000 
 50.000000 
 49.750000 
 50.000000 
 45.000000 
 2.00000 
 
 
 50% 
 65.000000 
 75.000000 
 70.000000 
 65.000000 
 70.000000 
 65.000000 
 3.00000 
 
 
 75% 
 80.000000 
 100.000000 
 90.000000 
 95.000000 
 90.000000 
 90.000000 
 5.00000 
 
 
 max 
 255.000000 
 190.000000 
 230.000000 
 194.000000 
 230.000000 
 180.000000 
 6.00000

### The first way to sum up column`s values using +

In [0]:
df['Total'] = df.HP + df.Attack + df.Defense + df['Sp. Atk']+ df['Sp. Def'] + df.Speed

df.sort_values(by = 'Total', ascending=False, inplace=True)
df.head()

Name 
 Type 1 
 Type 2 
 HP 
 Attack 
 Defense 
 Sp. Atk 
 Sp. Def 
 Speed 
 Generation 
 Legendary 
 Total 
 
 
 
 
 426 
 RayquazaMega Rayquaza 
 Dragon 
 Flying 
 105 
 180 
 100 
 180 
 100 
 115 
 3 
 True 
 780 
 
 
 164 
 MewtwoMega Mewtwo Y 
 Psychic 
 NaN 
 106 
 150 
 70 
 194 
 120 
 140 
 1 
 True 
 780 
 
 
 163 
 MewtwoMega Mewtwo X 
 Psychic 
 Fighting 
 106 
 190 
 100 
 154 
 100 
 130 
 1 
 True 
 780 
 
 
 422 
 KyogrePrimal Kyogre 
 Water 
 NaN 
 100 
 150 
 90 
 180 
 160 
 90 
 3 
 True 
 770 
 
 
 424 
 GroudonPrimal Groudon 
 Ground 
 Fire 
 100 
 180 
 160 
 150 
 90 
 90 
 3 
 True 
 770

**RayquazaMega Rayquaza** & **MewtwoMega Mewtwo X** & **MewtwoMega Mewtwo** Y atr the best! They all have 780 points!

### The second way to sum up column's values using iloc and slices

In [0]:
df['Total_iloc'] = df.iloc[:, 3:9].sum(axis = 1)
df.head()

Name 
 Type 1 
 Type 2 
 HP 
 Attack 
 Defense 
 Sp. Atk 
 Sp. Def 
 Speed 
 Generation 
 Legendary 
 Total 
 Total_iloc 
 
 
 
 
 426 
 RayquazaMega Rayquaza 
 Dragon 
 Flying 
 105 
 180 
 100 
 180 
 100 
 115 
 3 
 True 
 780 
 780 
 
 
 164 
 MewtwoMega Mewtwo Y 
 Psychic 
 NaN 
 106 
 150 
 70 
 194 
 120 
 140 
 1 
 True 
 780 
 780 
 
 
 163 
 MewtwoMega Mewtwo X 
 Psychic 
 Fighting 
 106 
 190 
 100 
 154 
 100 
 130 
 1 
 True 
 780 
 780 
 
 
 422 
 KyogrePrimal Kyogre 
 Water 
 NaN 
 100 
 150 
 90 
 180 
 160 
 90 
 3 
 True 
 770 
 770 
 
 
 424 
 GroudonPrimal Groudon 
 Ground 
 Fire 
 100 
 180 
 160 
 150 
 90 
 90 
 3 
 True 
 770 
 770

### Reordering columns

In [0]:
columns = list(df.columns)
columns

['Name',
 'Type 1',
 'Type 2',
 'HP',
 'Attack',
 'Defense',
 'Sp. Atk',
 'Sp. Def',
 'Speed',
 'Generation',
 'Legendary',
 'Total',
 'Total_iloc']

In [0]:
df = df[columns[0:3] + [columns[-2]] + columns[3:-2]]
df.head()

Name 
 Type 1 
 Type 2 
 Total 
 HP 
 Attack 
 Defense 
 Sp. Atk 
 Sp. Def 
 Speed 
 Generation 
 Legendary 
 
 
 
 
 426 
 RayquazaMega Rayquaza 
 Dragon 
 Flying 
 780 
 105 
 180 
 100 
 180 
 100 
 115 
 3 
 True 
 
 
 164 
 MewtwoMega Mewtwo Y 
 Psychic 
 NaN 
 780 
 106 
 150 
 70 
 194 
 120 
 140 
 1 
 True 
 
 
 163 
 MewtwoMega Mewtwo X 
 Psychic 
 Fighting 
 780 
 106 
 190 
 100 
 154 
 100 
 130 
 1 
 True 
 
 
 422 
 KyogrePrimal Kyogre 
 Water 
 NaN 
 770 
 100 
 150 
 90 
 180 
 160 
 90 
 3 
 True 
 
 
 424 
 GroudonPrimal Groudon 
 Ground 
 Fire 
 770 
 100 
 180 
 160 
 150 
 90 
 90 
 3 
 True

## Filtering data

In [0]:
df_grass = df[df['Type 1'] == 'Grass']
df_grass

Name 
 Type 1 
 Type 2 
 Total 
 HP 
 Attack 
 Defense 
 Sp. Atk 
 Sp. Def 
 Speed 
 Generation 
 Legendary 
 
 
 
 
 275 
 SceptileMega Sceptile 
 Grass 
 Dragon 
 630 
 70 
 110 
 75 
 145 
 85 
 145 
 3 
 False 
 
 
 3 
 VenusaurMega Venusaur 
 Grass 
 Poison 
 625 
 80 
 100 
 123 
 122 
 120 
 80 
 1 
 False 
 
 
 551 
 ShayminSky Forme 
 Grass 
 Flying 
 600 
 100 
 103 
 75 
 120 
 75 
 127 
 4 
 True 
 
 
 550 
 ShayminLand Forme 
 Grass 
 NaN 
 600 
 100 
 100 
 100 
 100 
 100 
 100 
 4 
 True 
 
 
 511 
 AbomasnowMega Abomasnow 
 Grass 
 Ice 
 594 
 90 
 132 
 105 
 132 
 105 
 30 
 4 
 False 
 
 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 
 
 451 
 Budew 
 Grass 
 Poison 
 280 
 40 
 30 
 35 
 50 
 70 
 55 
 4 
 False 
 
 
 467 
 Cherubi 
 Grass 
 NaN 
 275 
 45 
 35 
 45 
 62 
 53 
 35 
 4 
 False 
 
 
 202 
 Hoppip 
 Grass 
 Flying 
 250 
 35 
 35 
 40 
 35 
 55 
 50 
 2 
 False 
 
 
 296 
 Seedot 
 Grass 
 NaN 
 220 
 40 
 40 
 50 
 30 
 30 
 30 
 3 
 False 
 
 
 206 
 Sunkern 
 Grass 
 NaN 
 180 
 30 
 30 
 30 
 30 
 30 
 30 
 2 
 False 
 
 
 
 70 rows × 12 columns

In [0]:
df_grass['Type 1'].unique()

array(['Grass'], dtype=object)

In [0]:
df_grass_water = df[(df['Type 1'] == 'Grass') | (df['Type 1'] == 'Water')]
df_grass_water

Name 
 Type 1 
 Type 2 
 Total 
 HP 
 Attack 
 Defense 
 Sp. Atk 
 Sp. Def 
 Speed 
 Generation 
 Legendary 
 
 
 
 
 422 
 KyogrePrimal Kyogre 
 Water 
 NaN 
 770 
 100 
 150 
 90 
 180 
 160 
 90 
 3 
 True 
 
 
 541 
 Palkia 
 Water 
 Dragon 
 680 
 90 
 120 
 100 
 150 
 120 
 100 
 4 
 True 
 
 
 421 
 Kyogre 
 Water 
 NaN 
 670 
 100 
 100 
 90 
 150 
 140 
 90 
 3 
 True 
 
 
 141 
 GyaradosMega Gyarados 
 Water 
 Dark 
 640 
 95 
 155 
 109 
 70 
 130 
 81 
 1 
 False 
 
 
 283 
 SwampertMega Swampert 
 Water 
 Ground 
 635 
 100 
 150 
 110 
 95 
 110 
 70 
 3 
 False 
 
 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 
 
 296 
 Seedot 
 Grass 
 NaN 
 220 
 40 
 40 
 50 
 30 
 30 
 30 
 3 
 False 
 
 
 209 
 Wooper 
 Water 
 Ground 
 210 
 55 
 45 
 45 
 25 
 25 
 15 
 2 
 False 
 
 
 139 
 Magikarp 
 Water 
 NaN 
 200 
 20 
 10 
 55 
 15 
 20 
 80 
 1 
 False 
 
 
 381 
 Feebas 
 Water 
 NaN 
 200 
 20 
 15 
 20 
 10 
 55 
 80 
 3 
 False 
 
 
 206 
 Sunkern 
 Grass 
 NaN 
 180 
 30 
 30 
 30 
 30 
 30 
 30 
 2 
 False 
 
 
 
 182 rows × 12 columns

In [0]:
df_grass_water['Type 1'].unique()

array(['Water', 'Grass'], dtype=object)

### Replace idex

In [0]:
df_grass_water.reset_index()

index 
 Name 
 Type 1 
 Type 2 
 Total 
 HP 
 Attack 
 Defense 
 Sp. Atk 
 Sp. Def 
 Speed 
 Generation 
 Legendary 
 
 
 
 
 0 
 422 
 KyogrePrimal Kyogre 
 Water 
 NaN 
 770 
 100 
 150 
 90 
 180 
 160 
 90 
 3 
 True 
 
 
 1 
 541 
 Palkia 
 Water 
 Dragon 
 680 
 90 
 120 
 100 
 150 
 120 
 100 
 4 
 True 
 
 
 2 
 421 
 Kyogre 
 Water 
 NaN 
 670 
 100 
 100 
 90 
 150 
 140 
 90 
 3 
 True 
 
 
 3 
 141 
 GyaradosMega Gyarados 
 Water 
 Dark 
 640 
 95 
 155 
 109 
 70 
 130 
 81 
 1 
 False 
 
 
 4 
 283 
 SwampertMega Swampert 
 Water 
 Ground 
 635 
 100 
 150 
 110 
 95 
 110 
 70 
 3 
 False 
 
 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 
 
 177 
 296 
 Seedot 
 Grass 
 NaN 
 220 
 40 
 40 
 50 
 30 
 30 
 30 
 3 
 False 
 
 
 178 
 209 
 Wooper 
 Water 
 Ground 
 210 
 55 
 45 
 45 
 25 
 25 
 15 
 2 
 False 
 
 
 179 
 139 
 Magikarp 
 Water 
 NaN 
 200 
 20 
 10 
 55 
 15 
 20 
 80 
 1 
 False 
 
 
 180 
 381 
 Feebas 
 Water 
 NaN 
 200 
 20 
 15 
 20 
 10 
 55 
 80 
 3 
 False 
 
 
 181 
 206 
 Sunkern 
 Grass 
 NaN 
 180 
 30 
 30 
 30 
 30 
 30 
 30 
 2 
 False 
 
 
 
 182 rows × 13 columns

*reset_index()* - reset index and creates neww column with index value. New index column is created automatically.

In [0]:
df_grass.reset_index(drop = True)

Name 
 Type 1 
 Type 2 
 Total 
 HP 
 Attack 
 Defense 
 Sp. Atk 
 Sp. Def 
 Speed 
 Generation 
 Legendary 
 
 
 
 
 0 
 SceptileMega Sceptile 
 Grass 
 Dragon 
 630 
 70 
 110 
 75 
 145 
 85 
 145 
 3 
 False 
 
 
 1 
 VenusaurMega Venusaur 
 Grass 
 Poison 
 625 
 80 
 100 
 123 
 122 
 120 
 80 
 1 
 False 
 
 
 2 
 ShayminSky Forme 
 Grass 
 Flying 
 600 
 100 
 103 
 75 
 120 
 75 
 127 
 4 
 True 
 
 
 3 
 ShayminLand Forme 
 Grass 
 NaN 
 600 
 100 
 100 
 100 
 100 
 100 
 100 
 4 
 True 
 
 
 4 
 AbomasnowMega Abomasnow 
 Grass 
 Ice 
 594 
 90 
 132 
 105 
 132 
 105 
 30 
 4 
 False 
 
 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 
 
 65 
 Budew 
 Grass 
 Poison 
 280 
 40 
 30 
 35 
 50 
 70 
 55 
 4 
 False 
 
 
 66 
 Cherubi 
 Grass 
 NaN 
 275 
 45 
 35 
 45 
 62 
 53 
 35 
 4 
 False 
 
 
 67 
 Hoppip 
 Grass 
 Flying 
 250 
 35 
 35 
 40 
 35 
 55 
 50 
 2 
 False 
 
 
 68 
 Seedot 
 Grass 
 NaN 
 220 
 40 
 40 
 50 
 30 
 30 
 30 
 3 
 False 
 
 
 69 
 Sunkern 
 Grass 
 NaN 
 180 
 30 
 30 
 30 
 30 
 30 
 30 
 2 
 False 
 
 
 
 70 rows × 12 columns

*reset_index()* - resets index, deletes the previous column with old indexes and creates new column with index value. 

### Regex usage

*re.I* means ignore capitalization

In [0]:
import re

In [0]:
df[df['Type 1'].str.contains('fire|water', flags = re.I,  regex = True)]

Name 
 Type 1 
 Type 2 
 Total 
 HP 
 Attack 
 Defense 
 Sp. Atk 
 Sp. Def 
 Speed 
 Generation 
 Legendary 
 
 
 
 
 422 
 KyogrePrimal Kyogre 
 Water 
 NaN 
 770 
 100 
 150 
 90 
 180 
 160 
 90 
 3 
 True 
 
 
 541 
 Palkia 
 Water 
 Dragon 
 680 
 90 
 120 
 100 
 150 
 120 
 100 
 4 
 True 
 
 
 270 
 Ho-oh 
 Fire 
 Flying 
 680 
 106 
 130 
 90 
 110 
 154 
 90 
 2 
 True 
 
 
 421 
 Kyogre 
 Water 
 NaN 
 670 
 100 
 100 
 90 
 150 
 140 
 90 
 3 
 True 
 
 
 141 
 GyaradosMega Gyarados 
 Water 
 Dark 
 640 
 95 
 155 
 109 
 70 
 130 
 81 
 1 
 False 
 
 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 
 
 236 
 Slugma 
 Fire 
 NaN 
 250 
 40 
 40 
 40 
 70 
 40 
 20 
 2 
 False 
 
 
 293 
 Lotad 
 Water 
 Grass 
 220 
 40 
 30 
 30 
 40 
 50 
 30 
 3 
 False 
 
 
 209 
 Wooper 
 Water 
 Ground 
 210 
 55 
 45 
 45 
 25 
 25 
 15 
 2 
 False 
 
 
 139 
 Magikarp 
 Water 
 NaN 
 200 
 20 
 10 
 55 
 15 
 20 
 80 
 1 
 False 
 
 
 381 
 Feebas 
 Water 
 NaN 
 200 
 20 
 15 
 20 
 10 
 55 
 80 
 3 
 False 
 
 
 
 164 rows × 12 columns

In [0]:
df[df['Name'].str.contains('^pi*', regex = True, flags = re.I)]['Name'].unique().tolist()

['Palkia',
 'PinsirMega Pinsir',
 'PidgeotMega Pidgeot',
 'Porygon-Z',
 'Probopass',
 'Porygon2',
 'Poliwrath',
 'Pyroar',
 'Politoed',
 'Pinsir',
 'Pangoro',
 'Phione',
 'Pidgeot',
 'Primeape',
 'Purugly',
 'Piloswine',
 'Persian',
 'Pelipper',
 'Pignite',
 'Ponyta',
 'Pupitar',
 'Pachirisu',
 'Prinplup',
 'Plusle',
 'Parasect',
 'Porygon',
 'Poliwhirl',
 'Palpitoad',
 'Pidgeotto',
 'Pancham',
 'Pawniard',
 'PumpkabooSmall Size',
 'PumpkabooLarge Size',
 'PumpkabooAverage Size',
 'PumpkabooSuper Size',
 'Phanpy',
 'Psyduck',
 'Pikachu',
 'Panpour',
 'Pansear',
 'Pansage',
 'Piplup',
 'Phantump',
 'Poliwag',
 'Pineco',
 'Paras',
 'Purrloin',
 'Petilil',
 'Pidove',
 'Patrat',
 'Pidgey',
 'Poochyena',
 'Pichu']

## Conditional Changes


!!!!! Use of loc is important

In [0]:
df.loc[df['Type 1'] == 'Fire', 'Type 1'] = 'Flamer'

In [0]:
df[df['Type 1'] == 'Flamer']

Name 
 Type 1 
 Type 2 
 Total 
 HP 
 Attack 
 Defense 
 Sp. Atk 
 Sp. Def 
 Speed 
 Generation 
 Legendary 
 
 
 
 
 270 
 Ho-oh 
 Flamer 
 Flying 
 680 
 106 
 130 
 90 
 110 
 154 
 90 
 2 
 True 
 
 
 7 
 CharizardMega Charizard X 
 Flamer 
 Dragon 
 634 
 78 
 130 
 111 
 130 
 85 
 100 
 1 
 False 
 
 
 8 
 CharizardMega Charizard Y 
 Flamer 
 Flying 
 634 
 78 
 104 
 78 
 159 
 115 
 100 
 1 
 False 
 
 
 279 
 BlazikenMega Blaziken 
 Flamer 
 Fighting 
 630 
 80 
 160 
 80 
 130 
 80 
 100 
 3 
 False 
 
 
 799 
 Volcanion 
 Flamer 
 Water 
 600 
 80 
 110 
 120 
 130 
 90 
 70 
 6 
 True 
 
 
 542 
 Heatran 
 Flamer 
 Steel 
 600 
 91 
 90 
 106 
 130 
 106 
 77 
 4 
 True 
 
 
 158 
 Moltres 
 Flamer 
 Flying 
 580 
 90 
 100 
 90 
 125 
 85 
 90 
 1 
 True 
 
 
 263 
 Entei 
 Flamer 
 NaN 
 580 
 115 
 115 
 85 
 90 
 75 
 100 
 2 
 True 
 
 
 354 
 CameruptMega Camerupt 
 Flamer 
 Ground 
 560 
 70 
 120 
 100 
 145 
 105 
 20 
 3 
 False 
 
 
 64 
 Arcanine 
 Flamer 
 NaN 
 555 
 90 
 110 
 80 
 100 
 80 
 95 
 1 
 False 
 
 
 616 
 DarmanitanZen Mode 
 Flamer 
 Psychic 
 540 
 105 
 30 
 105 
 140 
 105 
 55 
 5 
 False 
 
 
 518 
 Magmortar 
 Flamer 
 NaN 
 540 
 75 
 95 
 67 
 125 
 95 
 83 
 4 
 False 
 
 
 437 
 Infernape 
 Flamer 
 Fighting 
 534 
 76 
 104 
 71 
 104 
 71 
 108 
 4 
 False 
 
 
 723 
 Delphox 
 Flamer 
 Psychic 
 534 
 75 
 69 
 72 
 114 
 100 
 104 
 6 
 False 
 
 
 6 
 Charizard 
 Flamer 
 Flying 
 534 
 78 
 84 
 78 
 109 
 85 
 100 
 1 
 False 
 
 
 171 
 Typhlosion 
 Flamer 
 NaN 
 534 
 78 
 84 
 78 
 109 
 85 
 100 
 2 
 False 
 
 
 278 
 Blaziken 
 Flamer 
 Fighting 
 530 
 80 
 120 
 70 
 110 
 70 
 80 
 3 
 False 
 
 
 559 
 Emboar 
 Flamer 
 Fighting 
 528 
 110 
 123 
 65 
 100 
 65 
 65 
 5 
 False 
 
 
 147 
 Flareon 
 Flamer 
 NaN 
 525 
 65 
 130 
 60 
 95 
 110 
 65 
 1 
 False 
 
 
 736 
 Pyroar 
 Flamer 
 Normal 
 507 
 86 
 68 
 72 
 109 
 66 
 106 
 6 
 False 
 
 
 43 
 Ninetales 
 Flamer 
 NaN 
 505 
 73 
 76 
 75 
 81 
 100 
 100 
 1 
 False 
 
 
 84 
 Rapidash 
 Flamer 
 NaN 
 500 
 65 
 100 
 70 
 80 
 80 
 105 
 1 
 False 
 
 
 731 
 Talonflame 
 Flamer 
 Flying 
 499 
 78 
 81 
 71 
 74 
 69 
 126 
 6 
 False 
 
 
 573 
 Simisear 
 Flamer 
 NaN 
 498 
 75 
 98 
 63 
 98 
 63 
 101 
 5 
 False 
 
 
 135 
 Magmar 
 Flamer 
 NaN 
 495 
 65 
 95 
 57 
 100 
 85 
 93 
 1 
 False 
 
 
 692 
 Heatmor 
 Flamer 
 NaN 
 484 
 85 
 97 
 66 
 105 
 66 
 65 
 5 
 False 
 
 
 615 
 DarmanitanStandard Mode 
 Flamer 
 NaN 
 480 
 105 
 140 
 55 
 30 
 55 
 95 
 5 
 False 
 
 
 355 
 Torkoal 
 Flamer 
 NaN 
 470 
 70 
 85 
 140 
 85 
 70 
 20 
 3 
 False 
 
 
 353 
 Camerupt 
 Flamer 
 Ground 
 460 
 70 
 100 
 70 
 105 
 75 
 40 
 3 
 False 
 
 
 558 
 Pignite 
 Flamer 
 Fighting 
 418 
 90 
 93 
 55 
 70 
 55 
 55 
 5 
 False 
 
 
 83 
 Ponyta 
 Flamer 
 NaN 
 410 
 50 
 85 
 55 
 65 
 65 
 90 
 1 
 False 
 
 
 237 
 Magcargo 
 Flamer 
 Rock 
 410 
 50 
 50 
 120 
 80 
 80 
 30 
 2 
 False 
 
 
 722 
 Braixen 
 Flamer 
 NaN 
 409 
 59 
 59 
 58 
 90 
 70 
 73 
 6 
 False 
 
 
 277 
 Combusken 
 Flamer 
 Fighting 
 405 
 60 
 85 
 60 
 85 
 60 
 55 
 3 
 False 
 
 
 170 
 Quilava 
 Flamer 
 NaN 
 405 
 58 
 64 
 58 
 80 
 65 
 80 
 2 
 False 
 
 
 436 
 Monferno 
 Flamer 
 Fighting 
 405 
 64 
 78 
 52 
 78 
 52 
 81 
 4 
 False 
 
 
 5 
 Charmeleon 
 Flamer 
 NaN 
 405 
 58 
 64 
 58 
 80 
 65 
 80 
 1 
 False 
 
 
 730 
 Fletchinder 
 Flamer 
 Flying 
 382 
 62 
 73 
 55 
 56 
 52 
 84 
 6 
 False 
 
 
 735 
 Litleo 
 Flamer 
 Normal 
 369 
 62 
 50 
 58 
 73 
 54 
 72 
 6 
 False 
 
 
 259 
 Magby 
 Flamer 
 NaN 
 365 
 45 
 75 
 37 
 70 
 55 
 83 
 2 
 False 
 
 
 63 
 Growlithe 
 Flamer 
 NaN 
 350 
 55 
 70 
 45 
 70 
 50 
 60 
 1 
 False 
 
 
 572 
 Pansear 
 Flamer 
 NaN 
 316 
 50 
 53 
 48 
 53 
 48 
 64 
 5 
 False 
 
 
 614 
 Darumaka 
 Flamer 
 NaN 
 315 
 70 
 90 
 45 
 15 
 45 
 50 
 5 
 False 
 
 
 276 
 Torchic 
 Flamer 
 NaN 
 310 
 45 
 60 
 40 
 70 
 50 
 45 

In [0]:
df.loc[df['Type 1'] == 'Flamer', 'HP'] = -10

In [0]:
df[df['Type 1'] == 'Flamer']

Name 
 Type 1 
 Type 2 
 Total 
 HP 
 Attack 
 Defense 
 Sp. Atk 
 Sp. Def 
 Speed 
 Generation 
 Legendary 
 
 
 
 
 270 
 Ho-oh 
 Flamer 
 Flying 
 680 
 -10 
 130 
 90 
 110 
 154 
 90 
 2 
 True 
 
 
 7 
 CharizardMega Charizard X 
 Flamer 
 Dragon 
 634 
 -10 
 130 
 111 
 130 
 85 
 100 
 1 
 False 
 
 
 8 
 CharizardMega Charizard Y 
 Flamer 
 Flying 
 634 
 -10 
 104 
 78 
 159 
 115 
 100 
 1 
 False 
 
 
 279 
 BlazikenMega Blaziken 
 Flamer 
 Fighting 
 630 
 -10 
 160 
 80 
 130 
 80 
 100 
 3 
 False 
 
 
 799 
 Volcanion 
 Flamer 
 Water 
 600 
 -10 
 110 
 120 
 130 
 90 
 70 
 6 
 True 
 
 
 542 
 Heatran 
 Flamer 
 Steel 
 600 
 -10 
 90 
 106 
 130 
 106 
 77 
 4 
 True 
 
 
 158 
 Moltres 
 Flamer 
 Flying 
 580 
 -10 
 100 
 90 
 125 
 85 
 90 
 1 
 True 
 
 
 263 
 Entei 
 Flamer 
 NaN 
 580 
 -10 
 115 
 85 
 90 
 75 
 100 
 2 
 True 
 
 
 354 
 CameruptMega Camerupt 
 Flamer 
 Ground 
 560 
 -10 
 120 
 100 
 145 
 105 
 20 
 3 
 False 
 
 
 64 
 Arcanine 
 Flamer 
 NaN 
 555 
 -10 
 110 
 80 
 100 
 80 
 95 
 1 
 False 
 
 
 616 
 DarmanitanZen Mode 
 Flamer 
 Psychic 
 540 
 -10 
 30 
 105 
 140 
 105 
 55 
 5 
 False 
 
 
 518 
 Magmortar 
 Flamer 
 NaN 
 540 
 -10 
 95 
 67 
 125 
 95 
 83 
 4 
 False 
 
 
 437 
 Infernape 
 Flamer 
 Fighting 
 534 
 -10 
 104 
 71 
 104 
 71 
 108 
 4 
 False 
 
 
 723 
 Delphox 
 Flamer 
 Psychic 
 534 
 -10 
 69 
 72 
 114 
 100 
 104 
 6 
 False 
 
 
 6 
 Charizard 
 Flamer 
 Flying 
 534 
 -10 
 84 
 78 
 109 
 85 
 100 
 1 
 False 
 
 
 171 
 Typhlosion 
 Flamer 
 NaN 
 534 
 -10 
 84 
 78 
 109 
 85 
 100 
 2 
 False 
 
 
 278 
 Blaziken 
 Flamer 
 Fighting 
 530 
 -10 
 120 
 70 
 110 
 70 
 80 
 3 
 False 
 
 
 559 
 Emboar 
 Flamer 
 Fighting 
 528 
 -10 
 123 
 65 
 100 
 65 
 65 
 5 
 False 
 
 
 147 
 Flareon 
 Flamer 
 NaN 
 525 
 -10 
 130 
 60 
 95 
 110 
 65 
 1 
 False 
 
 
 736 
 Pyroar 
 Flamer 
 Normal 
 507 
 -10 
 68 
 72 
 109 
 66 
 106 
 6 
 False 
 
 
 43 
 Ninetales 
 Flamer 
 NaN 
 505 
 -10 
 76 
 75 
 81 
 100 
 100 
 1 
 False 
 
 
 84 
 Rapidash 
 Flamer 
 NaN 
 500 
 -10 
 100 
 70 
 80 
 80 
 105 
 1 
 False 
 
 
 731 
 Talonflame 
 Flamer 
 Flying 
 499 
 -10 
 81 
 71 
 74 
 69 
 126 
 6 
 False 
 
 
 573 
 Simisear 
 Flamer 
 NaN 
 498 
 -10 
 98 
 63 
 98 
 63 
 101 
 5 
 False 
 
 
 135 
 Magmar 
 Flamer 
 NaN 
 495 
 -10 
 95 
 57 
 100 
 85 
 93 
 1 
 False 
 
 
 692 
 Heatmor 
 Flamer 
 NaN 
 484 
 -10 
 97 
 66 
 105 
 66 
 65 
 5 
 False 
 
 
 615 
 DarmanitanStandard Mode 
 Flamer 
 NaN 
 480 
 -10 
 140 
 55 
 30 
 55 
 95 
 5 
 False 
 
 
 355 
 Torkoal 
 Flamer 
 NaN 
 470 
 -10 
 85 
 140 
 85 
 70 
 20 
 3 
 False 
 
 
 353 
 Camerupt 
 Flamer 
 Ground 
 460 
 -10 
 100 
 70 
 105 
 75 
 40 
 3 
 False 
 
 
 558 
 Pignite 
 Flamer 
 Fighting 
 418 
 -10 
 93 
 55 
 70 
 55 
 55 
 5 
 False 
 
 
 83 
 Ponyta 
 Flamer 
 NaN 
 410 
 -10 
 85 
 55 
 65 
 65 
 90 
 1 
 False 
 
 
 237 
 Magcargo 
 Flamer 
 Rock 
 410 
 -10 
 50 
 120 
 80 
 80 
 30 
 2 
 False 
 
 
 722 
 Braixen 
 Flamer 
 NaN 
 409 
 -10 
 59 
 58 
 90 
 70 
 73 
 6 
 False 
 
 
 277 
 Combusken 
 Flamer 
 Fighting 
 405 
 -10 
 85 
 60 
 85 
 60 
 55 
 3 
 False 
 
 
 170 
 Quilava 
 Flamer 
 NaN 
 405 
 -10 
 64 
 58 
 80 
 65 
 80 
 2 
 False 
 
 
 436 
 Monferno 
 Flamer 
 Fighting 
 405 
 -10 
 78 
 52 
 78 
 52 
 81 
 4 
 False 
 
 
 5 
 Charmeleon 
 Flamer 
 NaN 
 405 
 -10 
 64 
 58 
 80 
 65 
 80 
 1 
 False 
 
 
 730 
 Fletchinder 
 Flamer 
 Flying 
 382 
 -10 
 73 
 55 
 56 
 52 
 84 
 6 
 False 
 
 
 735 
 Litleo 
 Flamer 
 Normal 
 369 
 -10 
 50 
 58 
 73 
 54 
 72 
 6 
 False 
 
 
 259 
 Magby 
 Flamer 
 NaN 
 365 
 -10 
 75 
 37 
 70 
 55 
 83 
 2 
 False 
 
 
 63 
 Growlithe 
 Flamer 
 NaN 
 350 
 -10 
 70 
 45 
 70 
 50 
 60 
 1 
 False 
 
 
 572 
 Pansear 
 Flamer 
 NaN 
 316 
 -10 
 53 
 48 
 53 
 48 
 64 
 5 
 False 
 
 
 614 
 Darumaka 
 Flamer 
 NaN 
 315 
 -10 
 90 
 45 
 15 
 45 
 50 
 5 
 False 
 
 
 276 
 Torchic 
 Flamer 
 Na

In [0]:
df.loc[df.HP < 100, ['Type 1', 'Type 2']] = ['Test1', 'Test2']

In [0]:
df[df.HP < 100]

Name 
 Type 1 
 Type 2 
 Total 
 HP 
 Attack 
 Defense 
 Sp. Atk 
 Sp. Def 
 Speed 
 Generation 
 Legendary 
 
 
 
 
 409 
 SalamenceMega Salamence 
 Test1 
 Test2 
 700 
 95 
 145 
 130 
 120 
 90 
 120 
 3 
 False 
 
 
 413 
 MetagrossMega Metagross 
 Test1 
 Test2 
 700 
 80 
 145 
 150 
 105 
 110 
 110 
 3 
 False 
 
 
 418 
 LatiasMega Latias 
 Test1 
 Test2 
 700 
 80 
 100 
 120 
 140 
 150 
 110 
 3 
 True 
 
 
 420 
 LatiosMega Latios 
 Test1 
 Test2 
 700 
 80 
 130 
 100 
 160 
 120 
 110 
 3 
 True 
 
 
 796 
 DiancieMega Diancie 
 Test1 
 Test2 
 700 
 50 
 160 
 110 
 160 
 110 
 110 
 6 
 True 
 
 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 
 
 13 
 Caterpie 
 Test1 
 Test2 
 195 
 45 
 30 
 35 
 20 
 20 
 45 
 1 
 False 
 
 
 288 
 Wurmple 
 Test1 
 Test2 
 195 
 45 
 45 
 35 
 20 
 30 
 20 
 3 
 False 
 
 
 446 
 Kricketot 
 Test1 
 Test2 
 194 
 37 
 25 
 41 
 25 
 41 
 25 
 4 
 False 
 
 
 322 
 Azurill 
 Test1 
 Test2 
 190 
 50 
 20 
 40 
 20 
 40 
 20 
 3 
 False 
 
 
 206 
 Sunkern 
 Test1 
 Test2 
 180 
 30 
 30 
 30 
 30 
 30 
 30 
 2 
 False 
 
 
 
 706 rows × 12 columns

## Aggregate statistics

In [0]:
df = df[(df['Type 1'] != 'Test1') & (df['Type 2'] != 'Test2')]

In [0]:
# check that Type 1 = Test1 was deleted
df['Type 1'].unique()

array(['Dragon', 'Psychic', 'Water', 'Ground', 'Normal', 'Rock', 'Ghost', 'Steel', 'Fairy', 'Dark', 'Grass', 'Ice', 'Fighting', 'Poison'], dtype=object)

In [0]:
df.groupby('Type 1').mean().sort_values('Defense', ascending = False)

Total 
 HP 
 Attack 
 Defense 
 Sp. Atk 
 Sp. Def 
 Speed 
 Generation 
 Legendary 
 
 
 Type 1 
 
 
 
 
 
 
 
 
 
 
 
 
 
 Ground 
 566.285714 
 106.714286 
 130.428571 
 116.000000 
 78.428571 
 73.714286 
 61.000000 
 3.571429 
 0.285714 
 
 
 Rock 
 607.000000 
 107.666667 
 125.000000 
 110.666667 
 96.333333 
 104.000000 
 63.333333 
 3.333333 
 0.000000 
 
 
 Steel 
 640.000000 
 100.000000 
 110.000000 
 110.000000 
 125.000000 
 100.000000 
 95.000000 
 3.500000 
 1.000000 
 
 
 Dragon 
 678.000000 
 110.900000 
 142.000000 
 102.100000 
 130.100000 
 96.500000 
 96.400000 
 4.500000 
 0.800000 
 
 
 Psychic 
 609.272727 
 114.545455 
 96.636364 
 93.454545 
 112.000000 
 102.909091 
 89.727273 
 2.636364 
 0.454545 
 
 
 Ghost 
 619.333333 
 150.000000 
 100.000000 
 88.000000 
 103.333333 
 91.333333 
 86.666667 
 4.000000 
 0.666667 
 
 
 Grass 
 546.000000 
 107.400000 
 97.600000 
 86.400000 
 102.400000 
 77.200000 
 75.000000 
 4.600000 
 0.400000 
 
 
 Dark 
 565.000000 
 112.000000 
 107.000000 
 84.000000 
 97.000000 
 81.666667 
 83.333333 
 5.000000 
 0.333333 
 
 
 Fairy 
 571.000000 
 113.500000 
 101.500000 
 83.500000 
 115.000000 
 93.500000 
 64.000000 
 6.000000 
 0.500000 
 
 
 Ice 
 503.333333 
 106.666667 
 103.333333 
 83.333333 
 75.000000 
 70.000000 
 65.000000 
 3.000000 
 0.000000 
 
 
 Water 
 528.722222 
 115.333333 
 88.000000 
 80.500000 
 89.666667 
 88.000000 
 67.222222 
 3.055556 
 0.166667 
 
 
 Fighting 
 481.333333 
 123.000000 
 120.000000 
 80.000000 
 41.666667 
 70.000000 
 46.666667 
 4.333333 
 0.000000 
 
 
 Poison 
 482.000000 
 102.666667 
 90.333333 
 75.000000 
 69.666667 
 81.333333 
 63.000000 
 2.666667 
 0.000000 
 
 
 Normal 
 502.238095 
 126.904762 
 82.571429 
 68.190476 
 69.952381 
 89.095238 
 65.523810 
 3.000000 
 0.095238

## Working with large amount of data

In [0]:
df_temp = pd.DataFrame(columns = df.columns[4:])

# devide .csv on chuncks

for df_i in pd.read_csv('pokemon_data.csv', chunksize = 10 ):
    result = df_i.groupby('Type 1').mean()

    df_temp = pd.concat([df_temp, result])

df_temp

HP 
 Attack 
 Defense 
 Sp. Atk 
 Sp. Def 
 Speed 
 Generation 
 Legendary 
 # 
 Type 2 
 
 
 
 
 Fire 
 66.200000 
 86.800000 
 73.600000 
 107.600000 
 80.000000 
 89.000000 
 1.0 
 False 
 5.400000 
 NaN 
 
 
 Grass 
 66.250000 
 73.250000 
 79.500000 
 91.750000 
 91.250000 
 66.250000 
 1.0 
 False 
 2.250000 
 NaN 
 
 
 Water 
 44.000000 
 48.000000 
 65.000000 
 50.000000 
 64.000000 
 43.000000 
 1.0 
 False 
 7.000000 
 NaN 
 
 
 Bug 
 52.857143 
 56.428571 
 42.857143 
 34.285714 
 47.142857 
 64.285714 
 1.0 
 False 
 12.857143 
 NaN 
 
 
 Water 
 72.333333 
 83.000000 
 100.000000 
 95.000000 
 100.000000 
 71.333333 
 1.0 
 False 
 8.666667 
 NaN 
 
 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 
 
 Fairy 
 126.000000 
 131.000000 
 95.000000 
 131.000000 
 98.000000 
 99.000000 
 6.0 
 True 
 716.000000 
 NaN 
 
 
 Fire 
 80.000000 
 110.000000 
 120.000000 
 130.000000 
 90.000000 
 70.000000 
 6.0 
 True 
 721.000000 
 NaN 
 
 
 Flying 
 62.500000 
 50.000000 
 57.500000 
 71.000000 
 60.000000 
 89.000000 
 6.0 
 False 
 714.500000 
 NaN 
 
 
 Psychic 
 80.000000 
 135.000000 
 60.000000 
 160.000000 
 130.000000 
 75.000000 
 6.0 
 True 
 720.000000 
 NaN 
 
 
 Rock 
 50.000000 
 130.000000 
 130.000000 
 130.000000 
 130.000000 
 80.000000 
 6.0 
 True 
 719.000000 
 NaN 
 
 
 
 361 rows × 10 columns